File: Animations/app/src/main/java/com/example/animations/MainActivity.java
package com.example.animations;

import android.graphics.ImageDecoder;
import android.graphics.drawable.Animatable;
import android.graphics.drawable.AnimatedImageDrawable;
import android.graphics.drawable.Drawable;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.View;
import android.widget.ImageView;

import java.io.IOException;

public class MainActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        final ImageView imageView = findViewById(R.id.animated_image_view);

        /*new Thread(new Runnable() {
            @Override
            public void run() {
                try {
                    final Drawable drawable1 = ImageDecoder.decodeDrawable(ImageDecoder.createSource(getResources(), R.drawable.android_project));

                    runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                            imageView.setImageDrawable(drawable1);
                        }
                    });
                    if (drawable1 instanceof Animatable){
                        ((AnimatedImageDrawable)drawable1).start();

                    }
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        }).start();*/

        // wolverine animation
       /* imageView.setImageDrawable(getDrawable(R.drawable.animation_playground));
        final Drawable drawable = imageView.getDrawable();
        if (drawable instanceof Animatable){
            ((Animatable)drawable).start();
        }*/

       //todo getting a bug
        imageView.setImageDrawable(getDrawable(R.drawable.avd_anim));
        final Drawable drawable = (imageView).getDrawable();
        if (drawable instanceof Animatable) {
            ((Animatable) drawable).start();
        }

        imageView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                imageView.setImageDrawable(getDrawable(R.drawable.avd_anim));
                final Drawable drawable = imageView.getDrawable();
                if (drawable instanceof Animatable) {
                    ((Animatable) drawable).start();
                }
            }
        });
    }
}

File: app/src/main/java/com/example/animated_images/MainActivity.kt
package com.example.animated_images

import android.annotation.SuppressLint
import android.graphics.ImageDecoder
import android.graphics.drawable.AnimatedImageDrawable
import android.graphics.drawable.AnimationDrawable
import android.os.Build
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import androidx.core.content.ContextCompat
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {


    val drawableList = listOf(R.drawable.angry_kid, R.drawable.monkey_animated)
    var pointer =0
    fun incrementPointer() {
        pointer ++
        if (pointer >= drawableList.size) {
            pointer = 0
        }

    }

    fun decrementPointer() {
        pointer --
        if (pointer < 0 ) {
            pointer = drawableList.size -1
        }

    }

    @SuppressLint("WrongThread")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)


        button_next.setOnClickListener{
            incrementPointer()
            animated_image.setImageDrawable(ContextCompat.getDrawable(this,drawableList[pointer]))


        }

        button_previous.setOnClickListener{
            decrementPointer()
            animated_image.setImageDrawable(ContextCompat.getDrawable(this,drawableList[pointer]))


        }

        button_play.setOnClickListener {
            when (pointer) {
                0 -> {animateGif(drawableList[pointer])}
                1-> {animateAnimationDrawable(drawableList[pointer])}

            }

        }


       // animateGif()

      //  animateAnimationDrawable()

    }

    private fun animateAnimationDrawable(id: Int) {
        val monkeyDrawable = ContextCompat.getDrawable(this, id)
        animated_image.setImageDrawable(monkeyDrawable)
        (monkeyDrawable as AnimationDrawable).start()
    }

    private fun animateGif(id: Int) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
            val angry_kid_drawable = ImageDecoder.decodeDrawable(ImageDecoder.createSource(resources,id))

        animated_image.setImageDrawable(angry_kid_drawable)

            (angry_kid_drawable as AnimatedImageDrawable).start()


        }
    }
}
